Update UI banned list

ptschip opened the pull request:
When a ban/unban happens internally (not through RPC) we need
to refresh the ban list in the UI.  Previous to this we
were only doing it if the ban was set through RPC only.

Commit:
Update UI banned list

-When a ban/unban happens internally (not through RPC) we need
to refresh the ban list in the UI.  Previous to this we
were only doing it if the ban was set through RPC only.

-Add 5 second Qtimer for automatically updating the banned list

When a ban comes to an end it will get deleted from the UI banned
list automatically rather than having to wait for another node
to be banned before the list would update.

-Keep use of SweepBanned() consistent for UI banned list

Dumping the banned list to disk every time the banned list is updated
is not necessary.  It gets dumped to disk every 15 mins and also when
the node shuts down.  Using sweepbanned is also consistent now with
the rest of the calls to bannning and unbanning of nodes.

-Several places in RPC calls for banning and unbanning
with redundant code.

-Make all calls to uiInterface.BannnedListChanged in src/net.cpp
This makes the code consistent for future use

-Add sweepbanned to listbanned

Justaphf commented on src/net.cpp:
```diff
@@ -546,6 +546,9 @@ bool CNode::IsBanned(CSubNet subnet)
     CSubNet subNet(addr);
     Ban(subNet, banReason, bantimeoffset, sinceUnixEpoch);
+
+    DumpBanlist(); //store banlist to disk
```
Redundant, will be called inside the subnet overload of the Ban function.

Justaphf commented on src/net.cpp:
```diff
@@ -563,11 +566,17 @@ void CNode::Ban(const CSubNet& subNet, const BanReason &banReason, int64_t banti
     CSubNet subNet(addr);
     return Unban(subNet);
+
+    DumpBanlist(); //store banlist to disk
```
Unreachable code, called after return statement (also redundant as should be called by the subnet overload of the Unband function).

Justaphf commented on src/net.cpp:
```diff
@@ -578,6 +587,9 @@ bool CNode::Unban(const CSubNet &subNet) {
     }
     return false;
+
+    DumpBanlist(); //store banlist to disk
```
Unreachable code, called after return statement.

ptschip commented:
thanks
i'll update

On 01/02/2017 5:13 AM, Justaphf wrote:
>
> *@Justaphf* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/net.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#pullrequestreview-19545575>:
>
> > @@ -578,6 +587,9 @@ bool CNode::Unban(const CSubNet &subNet) {
>          return true;
>      }
>      return false;
> +
> +    DumpBanlist(); //store banlist to disk
>
> Unreachable code, called after return statement.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#pullrequestreview-19545575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0D65lIyL3KT6audJS6XNInghfdH4ks5rYIUQgaJpZM4Lzz_L>.
>

ptschip commented:
ok, i've made the updates...have a look when you get some time.

On 01/02/2017 5:13 AM, Justaphf wrote:
>
> *@Justaphf* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/net.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#pullrequestreview-19545575>:
>
> > @@ -578,6 +587,9 @@ bool CNode::Unban(const CSubNet &subNet) {
>          return true;
>      }
>      return false;
> +
> +    DumpBanlist(); //store banlist to disk
>
> Unreachable code, called after return statement.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#pullrequestreview-19545575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0D65lIyL3KT6audJS6XNInghfdH4ks5rYIUQgaJpZM4Lzz_L>.
>

Justaphf commented:
Comments based just on review of the diffs, I can do a more in depth review tonight when I get home from work.  A couple other thoughts:
1. If we call the UI update explicitly from the ban/unban functions directly, I think the calls to update UI being made from within the RPC functions become redundant and could be removed.
2. I don't think we are automatically handling expiration of ban checks currently, only on-demand.  At the moment the only way expired bans get cleared is if another ban is added, the user manually unbans a node, the node restarts, the banned node attempts to connect and the ban is detected as expired, or (iirc, full code not in front of me) if SweepBanned is called (I don't recall the control paths which fire SweepBanned)
3. If we don't want to implement active monitoring to address point 2 (it may be too high effort-to-gain ratio), it would be useful to add a check and update in the `listbanned` RPC call such that that we make another on-demand check so that the RPC return is accurate at the time of the call (currently it is not if the only change in ban states is that an expiration of a ban occurred with no other ban/unban activity).

Justaphf commented on src/net.cpp:
```diff
@@ -574,6 +577,9 @@ bool CNode::Unban(const CSubNet &subNet) {
+        DumpBanlist(); //store banlist to disk
+        uiInterface.BannedListChanged();
+
         setBannedIsDirty = true;
```
~~Doesn't DumpBanList clear seBannedIsDirty?  I think this flag needs to be set prior to calling DumpBandList, not after.~~
I mis-rememebered, `DumpData` checks the dirty flag, `DumpBanList` does not use it at all (except it may get set indirectly through the call of `SweepBanned` if expired bans are detected and removed from the banmap).  I still think the flag should be set prior to the call to `DumpBanList`, but I think the usage of the `setBannedIsDirty` flag may be a larger issue because `DumpBanList` doesn't clear the flag which is only used to indicate that the file on disk and the banmap in memory are out of sync, which shouldn't be the case at the completion of writing to disk.

EDIT: Corrections after looking at implementation of DumpBanList, SweepBanned, and DumpData

ptschip commented:
On 01/02/2017 5:46 AM, Justaphf wrote:
>
> Comments based just on review of the diffs, I can do a more in depth
> review tonight when I get home from work. A couple other thoughts:
>
>  1. If we call the UI update explicitly from the ban/unban functions
>     directly, I think the calls to update UI being made from within
>     the RPC functions become redundant and could be removed.
>
hmm, good point..i'll take a look.
>
>  1. I don't think we are automatically handling expiration of ban
>     checks currently, only on-demand. At the moment the only way
>     expired bans get cleared is if another ban is added, the user
>     manually unbans a node, the node restarts, the banned node
>     attempts to connect and the ban is detected as expired, or (iirc,
>     full code not in front of me) if SweepBanned is called (I don't
>     recall the control paths which fire SweepBanned)
>

seems like a bit of effort and beyond the scope of this PR...but yes it
would be good to have.  It's only the nodes that get unbanned
though..and then those node may connect and clear the banned list
anway...so i think usefulness would be limited but again, good to have.
>
>  1. If we don't want to implement active monitoring to address point 2
>     (it may be too high effort-to-gain ratio), it would be useful to
>     add a check and update in the |listbanned| RPC call such that that
>     we make another on-demand check so that the RPC return is accurate
>     at the time of the call (currently it is not if the only change in
>     ban states is that an expiration of a ban occurred with no other
>     ban/unban activity).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#issuecomment-276660652>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0O3TF3ulleZNOPHyGAcbeGzCHBPsks5rYIy-gaJpZM4Lzz_L>.
>

ptschip commented:
no that's DumpData() you're thinking about...but i'll change it anyway
for consistency sake.

On 01/02/2017 5:49 AM, Justaphf wrote:
>
> *@Justaphf* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/net.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#pullrequestreview-19552007>:
>
> > @@ -574,6 +577,9 @@ bool CNode::Unban(const CSubNet &subNet) {
>      LOCK(cs_setBanned);
>      if (setBanned.erase(subNet))
>      {
> +        DumpBanlist(); //store banlist to disk
> +        uiInterface.BannedListChanged();
> +
>          setBannedIsDirty = true;
>
> Doesn't DumpBanList clear seBannedIsDirty? I think this flag needs to
> be set prior to calling DumpBandList, not after.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#pullrequestreview-19552007>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0NCqwlIx69R8pGf3o3JZ8v23DCKZks5rYI2DgaJpZM4Lzz_L>.
>

Justaphf commented:
Yea, I just edited my comment after realizing that.  I don't think the use of the `setBannedIsDirty` flag is consistent, it seems like that should only be used to indicate if the data on disk is out of sync with the banmap in memory, which shouldn't be the case since we just wrote to disk.  I'm thinking `DumpBanList` should clear that flag upon successful write to disk, provided it isn't being used elsewhere in the code to flag something different.

I'll dive into this more when I get home, review is much slower without the full dev environment in front of me.

ptschip commented:
as far as checking for bans when they expire, we could have some process
runs that checks every  so often...there's something in bitcoin which is
not really used which can launch processes and then shuts them down
properly when bitcoin shutsdown..the scheduler thread i believe its'
called.   I wonder though if there is something in QT we could
use...they must have something there that can update on occassion, some
sort of timer? 

On 01/02/2017 6:12 AM, Justaphf wrote:
>
> Yea, I just edited my comment after realizing that. I don't think the
> use of the |setBannedIsDirty| flag is consistent, it seems like that
> should only be used to indicate if the data on disk is out of sync
> with the banmap in memory, which shouldn't be the case since we just
> wrote to disk. I'm thinking |DumpBanList| should clear that flag upon
> successful write to disk, provided it isn't being used elsewhere in
> the code to flag something different.
>
> I'll dive into this more when I get home, review is much slower
> without the full dev environment in front of me.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#issuecomment-276666727>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0Infaz2ShEWNcgS5GvXLpp_U09ySks5rYJLhgaJpZM4Lzz_L>.
>

ptschip commented:
ok i found the timer,  QTimer..i'll see about adding another one...the
current timer updates every 250 milliseconds..probably too much for our
needs , i think every 5 seconds should be plenty.  I'll add anothe timer
and see how that goes.

On 01/02/2017 6:12 AM, Justaphf wrote:
>
> Yea, I just edited my comment after realizing that. I don't think the
> use of the |setBannedIsDirty| flag is consistent, it seems like that
> should only be used to indicate if the data on disk is out of sync
> with the banmap in memory, which shouldn't be the case since we just
> wrote to disk. I'm thinking |DumpBanList| should clear that flag upon
> successful write to disk, provided it isn't being used elsewhere in
> the code to flag something different.
>
> I'll dive into this more when I get home, review is much slower
> without the full dev environment in front of me.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#issuecomment-276666727>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0Infaz2ShEWNcgS5GvXLpp_U09ySks5rYJLhgaJpZM4Lzz_L>.
>

ptschip commented:
@Justaphf  I got the Qtimer to work...every 5 seconds the banned list in the UI will auto update.

Justaphf commented on src/net.cpp:
```diff
@@ -563,6 +563,9 @@ void CNode::Ban(const CSubNet& subNet, const BanReason &banReason, int64_t banti
 
     setBannedIsDirty = true;
+
+    DumpBanlist(); //store banlist to disk
```
Is it necessary to write to disk every time a ban comes in?  We can update the UI without writing to disk and having the `setBannedIsDirty` flag will catch it the next time `DumpData` gets hit (granted that's on a 15 minute interval).  My main concern here is what if we suddenly need to ban lots of nodes say during a DoS attack, we then end up writing to disk for each node banned (i.e. 50 bans = 50 writes where each write only adds one additional ban to the list).

Justaphf commented on src/net.cpp:
```diff
@@ -575,6 +578,9 @@ bool CNode::Unban(const CSubNet &subNet) {
     {
         setBannedIsDirty = true;
+
+        DumpBanlist(); //store banlist to disk
```
~~After looking through the code more, I think writing to disk for unbanning a node is unnecessary.  The only time we ever read this file from disk is when we start the node, which immediately calls `SweepBanned` to clear out expired bans.~~

Edit: Upon further review, there are only two control paths that actually call unban, both of which are user initiated (rpc/rpcconsole) and are prematurely ending a ban, so writing to file here makes perfect sense.

Justaphf commented on src/qt/guiconstants.h:
```diff
@@ -7,7 +7,8 @@
 /* Milliseconds between model updates */
-static const int MODEL_UPDATE_DELAY = 250;
+static const int MODEL_UPDATE_DELAY1 = 250;
+static const int MODEL_UPDATE_DELAY2 = 5000;
```
For readability, I think it would make more sense to name these constants `MODEL_UPDATE_DELAY_TPS` and `MODEL_UPDATE_DELAY_BAN`, since the original deals with updating the transactions per second and the new one deals with clearing expired bans.

Justaphf commented on src/qt/clientmodel.h:
```diff
@@ -108,7 +110,8 @@ class ClientModel : public QObject
 public Q_SLOTS:
-    void updateTimer();
+    void updateTimer1();
+    void updateTimer2();
```
For readability, I think it would make more sense to name these functions `updateTimer_TPS` and `updateTimer_Ban`, since the original deals with updating the transactions per second and the new one deals with clearing expired bans.

Justaphf commented on src/qt/clientmodel.h:
```diff
@@ -88,7 +88,8 @@ class ClientModel : public QObject
 
-    QTimer *pollTimer;
+    QTimer *pollTimer1;
+    QTimer *pollTimer2;
```
For readability, I think it would make more sense to name these timers `pollTimer_TPS` and `pollTimer_Ban`, since the original deals with updating the transactions per second and the new one deals with clearing expired bans.

Justaphf commented on src/qt/clientmodel.h:
```diff
@@ -99,7 +100,8 @@ class ClientModel : public QObject
     void bytesChanged(quint64 totalBytesIn, quint64 totalBytesOut);
-    void transactionsPerSecondChanged(double tansactionsPerSecond);  // BU:
+    void transactionsPerSecondChanged(double tansactionsPerSecond);
+    void updateBanlist2();
```
What is the purpose of `updateBanlist2`?  I don't see this used anywhere.

Justaphf commented on src/qt/clientmodel.cpp:
```diff
@@ -118,13 +122,24 @@ double ClientModel::getVerificationProgress(const CBlockIndex *tipIn) const
+    CNode::SweepBanned();
+
+    qDebug() << QString("%1: Requesting update for peer banlist").arg(__func__);
+    QMetaObject::invokeMethod(this, "updateBanlist", Qt::QueuedConnection);
```
With the exception of the call to `SweepBanned`, this exactly duplicates the function `BannedListChanged`.  I'm not very familiar with Qt programming, but would replacing those bottom two lines with `uiInterface.BannedListChanged();` have any issues?  This is consistent with usage in other areas of the application and improves code reuse by eliminating duplicate code.

Justaphf commented:
A couple of additional places where now redundant calls to update the UI and dump the ban list to disk:
1. qt/rpcconsole.cpp - `RPCConsole::banSelectedNode`
2. qt/rpcconsole.cpp - `RPCConsole::unbanSelectedNode`
3. rpc/net.cpp - `clearbanned`

For consistency we should consider updating `CNode::ClearBanned` to both dump to disk and update the UI, then remove the duplicate calls from rpc/net.cpp.  This is only ever called from one control path via RPC, but since the same functionality has been moved to CNode for ban/unban it would all be in the same location.

ptschip commented:
the first timer is used for 3 different updates, and the second for
only one right now but I already have in mind something else that i want
to use that timer for.  I think keeping it general is the way to go.

On 01/02/2017 10:17 PM, Justaphf wrote:
>
> *@Justaphf* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/qt/guiconstants.h
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#pullrequestreview-19719758>:
>
> > @@ -7,7 +7,8 @@
>  #define BITCOIN_QT_GUICONSTANTS_H
>  
>  /* Milliseconds between model updates */
> -static const int MODEL_UPDATE_DELAY = 250;
> +static const int MODEL_UPDATE_DELAY1 = 250;
> +static const int MODEL_UPDATE_DELAY2 = 5000;
>
> For readability, I think it would make more sense to name these
> constants |MODEL_UPDATE_DELAY_TPS| and |MODEL_UPDATE_DELAY_BAN|, since
> the original deals with updating the transactions per second and the
> new one deals with clearing expired bans.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#pullrequestreview-19719758>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0NSs4nEk6OnoQnPcpwvgMzKWi7riks5rYXTxgaJpZM4Lzz_L>.
>

ptschip commented:
sure , perhaps we don't need to do that, we can just do a sweepbanlist
and then when the node shuts down it will write the updates to disk.

On 01/02/2017 9:55 PM, Justaphf wrote:
>
> *@Justaphf* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/net.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#pullrequestreview-19718374>:
>
> > @@ -575,6 +578,9 @@ bool CNode::Unban(const CSubNet &subNet) {
>      if (setBanned.erase(subNet))
>      {
>          setBannedIsDirty = true;
> +
> +        DumpBanlist(); //store banlist to disk
>
> After looking through the code more, I think writing to disk for
> unbanning a node is unnecessary. The only time we ever read this file
> from disk is when we start the node, which immediately calls
> |SweepBanned| to clear out expired bans.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#pullrequestreview-19718374>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0MGcnAtbZHYFvr-zdw3E-NmC_zddks5rYW_igaJpZM4Lzz_L>.
>

ptschip commented:
@Justaphf thanks, i made the changes for consistency and tested them out.  I think it's a somewhat minor issue that we only use sweebanned() but it's good to be consistent and also it prevents anyone from somehow gaming the node by making us access the disk frequently...I suppose there is an attack vector there possible , although unlikely:   But it's good to plug the holes.  Thanks for the detailed review!

As for the naming convention around the polltimers, as I said, I think keeping them generic is the way to go since they can be used to call other methods.  The first polltimer already calls three diff methods so I think it's fine the way it is.

Justaphf commented:
@ptschip Great, I'll look more in-depth when I get home.  A couple of quick comments/questions from a glance at the latest code changes:
1. I'm OK with keeping the names for the timers & associated methods general, though I would still recommend making the names more distinguishable at a glance by using something more meaningful than 1 and 2.  In this case since no matter how many/which actions are tied to the timers, the interval value will remain the same. So what I would suggest to be a better distinguishing identifier than my original comment would be the interval itself (i.e. `MODEL_UPDATE_DELAY_250MS` and `MODEL_UPDATE_DELAY_5000MS`).
2. In rpc/net.cpp you don't need to call `CNode::SweepBanned` because `CNode::ClearBanned` has already emptied the banmap and there won't be anything to iterate over.
3. Suggest moving the UI update call from `clearbanned` in rpc/net.cpp to `CNode::ClearBanned` in net.cpp to keep the UI update calls consistently located wtih Ban/Unban.  This way if we, in the future, allow clear banned to be called elsewhere in the code, the UI updating is already handled.
4. In qt/rpcconsole.cpp the call to `CNode::SweepBanned` in `RPCConsole::unbanSelectedNode` is redundant because `CNode::Unban` calls it.  The only time this wouldn't be true is if the ban address didn't already exist in the banmap to be removed, but I would argue in this case we didn't need to call sweep anyway.
5. In qt/rpcconsole.cpp, isn't the call to `clientModel->getBanTableModel()->refresh()` in both `RPCConsole::unbanSelectedNode` and `RPCConsole::banSelectedNode` a redundant UI update, just via a different method?  I think it would be handled by the UI update calls now being made at the CNode level.
6. For both `CNode::Unban` and `CNode::ClearBanned`, I think it should actually be OK to call `DumpBanList`.  The only way to reach these control paths is by direct user interaction (RPC or UI) and by still writing to disk for these there is no net change to that functionality.  My concern regarding `CNode::Ban` was that it could be initiated by the client automatically based on the banning rules, not just by direct user interaction.  I was actually thinking about this more this morning and maybe it would make sense to modify the signature to `CNode::Ban(const CSubNet& subNet, const BanReason &banReason, int64_t bantimeoffset, bool sinceUnixEpoch, bool dumpBanList = false)` and just put a conditional call to `DumpBanList`
7. Do we need to update/add any unit/regression tests for these changes (or because they were missing to begin with).

Some of the items are probably outside the scope of this PR, but they are related and I wanted to mention them while I was thinking about them.

ptschip commented:
On 02/02/2017 6:54 AM, Justaphf wrote:
>
> @ptschip <https://github.com/ptschip> Great, I'll look more in-depth
> when I get home. A couple of quick comments/questions from a glance at
> the latest code changes:
>
>  1. I'm OK with keeping the names for the timers & associated methods
>     general, though I would still recommend making the names more
>     distinguishable at a glance by using something more meaningful
>     than 1 and 2. In this case since no matter how many/which actions
>     are tied to the timers, the interval value will remain the same.
>     So what I would suggest to be a better distinguishing identifier
>     than my original comment would be the interval itself (i.e.
>     |MODEL_UPDATE_DELAY_250MS| and |MODEL_UPDATE_DELAY_5000MS|).
>
i think it's better as is, we may want to tweak the values later and
then we don't have change the code so much.  It really doesn't take long
to figure out what the delays are for each timer by looking at the code.
>
>  1. In rpc/net.cpp you don't need to call |CNode::SweepBanned| because
>     |CNode::ClearBanned| has already emptied the banmap and there
>     won't be anything to iterate over.
>
good point...i'll change that.
>
>  1. Suggest moving the UI update call from |clearbanned| in
>     rpc/net.cpp to |CNode::ClearBanned| in net.cpp to keep the UI
>     update calls consistently located wtih Ban/Unban. This way if we,
>     in the future, allow clear banned to be called elsewhere in the
>     code, the UI updating is already handled.
>  2. In qt/rpcconsole.cpp the call to |CNode::SweepBanned| in
>     |RPCConsole::unbanSelectedNode| is redundant because
>     |CNode::Unban| calls it. The only time this wouldn't be true is if
>     the ban address didn't already exist in the banmap to be removed,
>     but I would argue in this case we didn't need to call sweep anyway.
>

true...changed.
>
>  1. In qt/rpcconsole.cpp, isn't the call to
>     |clientModel->getBanTableModel()->refresh()| in both
>     |RPCConsole::unbanSelectedNode| and |RPCConsole::banSelectedNode|
>     a redundant UI update, just via a different method? I think it
>     would be handled by the UI update calls now being made at the
>     CNode level.
>
we don't really need it there and we can just keep everything in
cnode::ban/unbun...i took them out and tested it and it works.
>
>  1. For both |CNode::Unban| and |CNode::ClearBanned|, I think it
>     should actually be OK to call |DumpBanList|. The only way to reach
>     these control paths is by direct user interaction (RPC or UI) and
>     by still writing to disk for these there is no net change to that
>     functionality. My concern regarding |CNode::Ban| was that it could
>     be initiated by the client automatically based on the banning
>     rules, not just by direct user interaction. I was actually
>     thinking about this more this morning and maybe it would make
>     sense to modify the signature to |CNode::Ban(const CSubNet&
>     subNet, const BanReason &banReason, int64_t bantimeoffset, bool
>     sinceUnixEpoch, bool dumpBanList = false)| and just put a
>     conditional call to |DumpBanList|
>
|i think that's beyond scope...maybe something you can follow up on later.|
>
>  1. Do we need to update/add any unit/regression tests for these
>     changes (or because they were missing to begin with).
>
there is already a suite of rpc tests for banning/unbanning...

> Some of the items are probably outside the scope of this PR, but they
> are related and I wanted to mention them while I was thinking about them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#issuecomment-276978914>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0N9lLiXPM72utblTZJJ_pIMRoJpiks5rYe4SgaJpZM4Lzz_L>.
>

Justaphf commented:
Any reason not to move the call to `uiInterface.BannedListChanged();` from `clearbanned` in rpc/net.cpp to `CNode::ClearBanned`?  With that move all ban sub-system related UI update triggers are centrally located in one place.

Other than that code review looks good.  I'll start implementation testing tomorrow.

On a side note, I saw that you've started removing the BU prefix on comments.  Is this a coding style change post 1.0?  Previously the suggested coding style was to use that prefix to denote BU modified code sections.

ptschip commented:
@Justaphf   Ok made that final change moving uiInterface.BannnedListChanges into cnode::clearbanned.

As for BU comments, we don't need to specify BU anymore, it's all BU now.

gandrewstone commented:
even though we are not rebasing from core anymore, lets leave the BU prefix comments in because we will still be cherry-picking features from them.

ptschip commented:
@gandrewstone OK, i put the BU comments back in and squashed everything down..this one is good to go.

gandrewstone commented:
@Justaphf you've been extremely active in reviewing this PR.  Can I get a formal "accept" from you?

Justaphf commented:
@gandrewstone @ptschip I've tested on Windows running both `bitcoin-qt` and `bitcoind`.  In general the UI updates on the timer are working as I expect they should, however I did have a few comments after implementation testing and another pass of code review, detailed below.  Point 1 I think should be addressed as part of this PR.  The others may be out of scope and needing their own separate PRs.

1. The 3 calls to `uiInterface.BannedListChanged()` in net.cpp are made while a lock to `cs_setBanned` is held.  I would recommend scoping the lock such that it is released before the UI update call is made.  I don't know that we will see any performance issues in this case, but I have found UI calls to be potentially slow so we shouldn't hold the lock while performing UI updates if we don't need to.

2. User initiated calls (via RPC or via direct UI interaction) to ban/unban/clearbanned no longer automatically cause an immediate update to the `banlist.dat` file via the `DumpBanned` method.  This is a loss of previous functionality and now the only time `banlist.dat` is updated is when the 15 minute timer hits to cause `DumpData` to fire (or on graceful node shutdown).  We discussed earlier in the review with regards to only calling `DumpBanned` when the action was initiated by direct user interaction as I could see a potential, although unlikely, issue if a lot of nodes get banned near simultaneously (potentially by the DoS mitigation ban rules) which could cause an excessive burst of disk writes.  My suggested fix is to update the signature of these 3 functions to take a flag to indicate if we should be writing to disk or not, however this may be beyond the scope of this PR and could have it's own follow-up PR.

3. Since the timer is in the Qt code, if you are running `bitcoind` instead of the desktop client, the issue of expired bans not being detected/removed until some external action causes `SweepBanned` to be called is not fixed.  You can only tell this if you call `listbanned` via RPC, so it is much less visible.  I think a simple fix would be to call `SweepBanned` inside of the handler for the `listbanned` RPC call.

4. When running `bitcoind`, I have also found that after the ban expires, the previously banned node may reconnect successfully, yet `listbanned` still shows the now reconnected node as banned (with the expired ban).  Simple test case was to take two of my local test nodes set to automatically connect to each other, on one node add a ban for 10 minutes for the other node, wait more than 10 minutes (i waited about 30 to also ensure I covered the `DumpData` 15 minute event to fire), call listbanned => see node still listed as banned, call getpeerinfo => see previously banned node is now connected, call listbanned => see the node still listed as banned.

ptschip commented:
@justaphf

On 09/02/2017 8:59 PM, Justaphf wrote:
>
> @gandrewstone <https://github.com/gandrewstone> @ptschip
> <https://github.com/ptschip> I've tested on Windows running both
> |bitcoin-qt| and |bitcoind|. In general the UI updates on the timer
> are working as I expect they should, however I did have a few comments
> after implementation testing and another pass of code review, detailed
> below. Point 1 I think should be addressed as part of this PR. The
> others may be out of scope and needing their own separate PRs.
>
> 1.
>
>     The 3 calls to |uiInterface.BannedListChanged()| in net.cpp are
>     made while a lock to |cs_setBanned| is held. I would recommend
>     scoping the lock such that it is released before the UI update
>     call is made. I don't know that we will see any performance issues
>     in this case, but I have found UI calls to be potentially slow so
>     we shouldn't hold the lock while performing UI updates if we don't
>     need to.
>
  they are only called when a ban or unban happens so i don't think this
is much a performance issue IMO. furthermore while 2 of the changes are
easily made the third would require more refactoring.
>
> 1.
>
>
> 2.
>
>     User initiated calls (via RPC or via direct UI interaction) to
>     ban/unban/clearbanned no longer automatically cause an immediate
>     update to the |banlist.dat| file via the |DumpBanned| method. This
>     is a loss of previous functionality and now the only time
>     |banlist.dat| is updated is when the 15 minute timer hits to cause
>     |DumpData| to fire (or on graceful node shutdown). We discussed
>     earlier in the review with regards to only calling |DumpBanned|
>     when the action was initiated by direct user interaction as I
>     could see a potential, although unlikely, issue if a lot of nodes
>     get banned near simultaneously (potentially by the DoS mitigation
>     ban rules) which could cause an excessive burst of disk writes. My
>     suggested fix is to update the signature of these 3 functions to
>     take a flag to indicate if we should be writing to disk or not,
>     however this may be beyond the scope of this PR and could have
>     it's own follow-up PR.
>
seems out of scope.
>
> 1.
>
>
> 2.
>
>     Since the timer is in the Qt code, if you are running |bitcoind|
>     instead of the desktop client, the issue of expired bans not being
>     detected/removed until some external action causes |SweepBanned|
>     to be called is not fixed. You can only tell this if you call
>     |listbanned| via RPC, so it is much less visible. I think a simple
>     fix would be to call |SweepBanned| inside of the handler for the
>     |listbanned| RPC call.
>
good catch...i'll update listbanned...
>
> 1.
>
>
> 2.
>
>     When running |bitcoind|, I have also found that after the ban
>     expires, the previously banned node may reconnect successfully,
>     yet |listbanned| still shows the now reconnected node as banned
>     (with the expired ban). Simple test case was to take two of my
>     local test nodes set to automatically connect to each other, on
>     one node add a ban for 10 minutes for the other node, wait more
>     than 10 minutes (i waited about 30 to also ensure I covered the
>     |DumpData| 15 minute event to fire), call listbanned => see node
>     still listed as banned, call getpeerinfo => see previously banned
>     node is now connected, call listbanned => see the node still
>     listed as banned.
>
same as above.
>
> 1.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#issuecomment-278858078>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0HIaQDNMAETonhApoQ8BcAjQzqtLks5ra-6YgaJpZM4Lzz_L>.
>

Commit:
Make CNode::DumpBanList local to net.cpp

It only gets called in net.cpp so no need to have
it in net.h

Remove redunant SweepBanned from CNode::DumpBanList
Remove redundant call to SweepBanned() from updateTimer2

ptschip commented:
@Justaphf Now that we added SweepBanned() to GetBanned() we can remove the SweepBanned() from DumpBanList() since it's now redundant.  see last commit.

Justaphf commented:
@ptschip I'll run implementation tests later today when I'm back home.  Based on code review things look good.  Only point I could make is that `DumpBanList` is no longer called anywhere outside of `net.cpp` so we could remove it from `net.h` and treat as local the way that `DumpAddresses` is treated.  Not sure if you want to include in this PR or if it should be a separate clean-up PR.

ptschip commented:
@Justaphf If we're doing these refactors we might as well take it out of net.h...makes sense. I'll do that and then when you get home if you can do a final run through, that would be great :)

Justaphf commented:
@ptschip It looks to me like adding `SweepBanned` to `GetBanned` also makes the call in `updateTimer2` redundant since the refresh of the ban table for the UI, in bantablemodel's `refreshBanlist`, calls `GetBanned`.  I'm going to test with that one line commented out and will let you know if everything still updates on the timer.

Justaphf commented:
I've run into issues building on windows with mingw now.  I think it has to do with the merge of 5bd2a94 into dev, but I'm getting errors in the execution of `./configure` now.  Spent most of the evening looking into that issue.  I'll try cross-compiling tomorrow so I can finish testing this PR.

Justaphf commented:
@ptschip I've tested and everything is working as expected now.  I also tried commenting out the call to `SweepBanned` in `updateTimer2` and the update of the ban table in the UI still works as expected so I would remove that redundant call and this should be good to merge.

ptschip commented:
@Justaphf thanks, done.

@gandrewstone looks like this one is ready.

gandrewstone commented:
Looks good to me.  But I have one question:  Is BannedListChanged thread safe?  It can now be called by the timer and via various APIs.  Unfortunately it just calls something else which calls something else (and so on) so I can't quickly determine whether it needs a lock or whether it is already thread safe.

ptschip commented:
@gandrewstone nice catch, yes you were right, we need a lock now...i added a lock on cs_setBanned for the BannedListChanged() calls.

Justaphf commented:
@ptschip @gandrewstone Perhaps I'm missing something coming from a C# background and not very familiar with Qt, but why is 'BannedListChanged' not thread safe?  Isn't the whole point of calling  `invokeMethod` to ensure synchronization of any UI related calls back to the UI thread?

ptschip commented:
you know actually i'm not sure now...but we're calling it from different
places now, not just one and I don't know enough about QT to be sure one
way or another, so putting an extra lock on cs_setBanned seems sensible
and is not an issue in terms of resources.

On 17/02/2017 5:36 PM, Justaphf wrote:
>
> @ptschip <https://github.com/ptschip> @gandrewstone
> <https://github.com/gandrewstone> Perhaps I'm missing something coming
> from a C# background and not very familiar with Qt, but why is
> 'BannedListChanged' not thread safe? Isn't the whole point of calling
> |invokeMethod| to ensure synchronization of any UI related calls back
> to the UI thread?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#issuecomment-280810774>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0OklXD64OjcKyBB2LXyDT7e3IPSjks5rdks3gaJpZM4Lzz_L>.
>

Justaphf commented:
I did some quick reading and it looks like since the `invokeMethod` works as I expected.  By using the `Qt::QueuedConnection` it is putting the call on the UI thread's message queue to be handled by the UI thread.  See https://books.google.com/books?id=t_3JCgAAQBAJ&pg=PT363&lpg=PT363&dq=Qt::QueuedConnection+invokeMethod&source=bl&ots=gyZ-eejCiE&sig=8oQ8K6jtoJJiVRftnRq2ENuXYMw&hl=en&sa=X&ved=0ahUKEwiJjp_UxZjSAhWmxYMKHWK5BbUQ6AEIQDAF#v=onepage&q=Qt%3A%3AQueuedConnection%20invokeMethod&f=false

ptschip commented:
@justaphf also, what's to prevent someone in the future from calling
BannedListChanged() directly.  But I think you're probably right about
the syncronization of the uiInteface calls but I still think it's
prudent to have the extra lock anyway IMO.

On 17/02/2017 5:36 PM, Justaphf wrote:
>
> @ptschip <https://github.com/ptschip> @gandrewstone
> <https://github.com/gandrewstone> Perhaps I'm missing something coming
> from a C# background and not very familiar with Qt, but why is
> 'BannedListChanged' not thread safe? Isn't the whole point of calling
> |invokeMethod| to ensure synchronization of any UI related calls back
> to the UI thread?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/262#issuecomment-280810774>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0OklXD64OjcKyBB2LXyDT7e3IPSjks5rdks3gaJpZM4Lzz_L>.
>

Justaphf commented:
I think putting the lock in `BannedListChanged` is the wrong level because all it does is put a message on the message queue to be processed by another thread.  To really add lock protection it would need to be in the method that's invoked, not the method doing the invoking.

ptschip commented:
@Justaphf you're right i'll change that , will push in a few mins.

ptschip commented:
@Justaphf   there is also something call layoutabouttobechanged() but i'm not sure it gives us the syncronization we're looking for, i think it's just used for updating the actual ui...i still feel uneasy without the lock, at least for now.

void BanTableModel::refresh()
{
    Q_EMIT layoutAboutToBeChanged();
    priv->refreshBanlist();
    Q_EMIT layoutChanged();
}

Justaphf commented:
I'm more confident in my previous statement now after having done some additional reading.  [QMetaObject Class Reference](http://ftp.ics.uci.edu/pub/centos0/ics-custom-build/BUILD/PyQt-x11-gpl-4.7.2/doc/html/qmetaobject.html#invokeMethod)

Since we are using `Qt::QueuedConnection`, this is a non-blocking asynchronous call and `BannedListChanged` can potentially return before `updateBanList` even begins processing.

gandrewstone commented:
ok I think @Justaphf is correct.  I just found another link that corrobrates his: http://doc.qt.io/qt-4.8/qmetaobject.html#invokeMethod
Specifically it says:
"If type is Qt::QueuedConnection, a QEvent will be sent and the member is invoked as soon as the application enters the main event loop."

Since QT's main event loop is running in a single thread so we are ok and its async anyway, with the passed flag, so the proposed lock won't protect the invoked method.  And I think that its fair to assume that invokeMethod itself must be thread safe since its job is to handle threading issues.

ptschip commented:
Ok, sounds good to me..i'll back that out and push up in minute...

ptschip commented:
ok done...

ptschip commented:
@gandrewstone np, it was a good learning experience in QT for us.

Justaphf commented:
@gandrewstone It's good to be sure, especially with potential threading issues.  Plus I learned quite a bit about Qt's threading in the process.
